### PR TITLE
[BEAM-3089] Fix default values in FlinkPipelineOptions / Add tests

### PR DESCRIPTION
**Pull request by mxm**
[BEAM-3089] Use Flink cluster parallelism if no parallelism provided

- The Flink Runner didn't pick up the parallelism from the Flink command-line if
provided there.

- A warning is printed if the parallelism can't be inferred and we fall back to
  parallelism of 1 (which is also current behavior).

- Local tests can now also use a parallelism > 1 which might help for test coverage.

- Further, tests verify as much from the current behavior as possible.

- [BEAM-3089] Add test for FlinkExecutionEnvironments
- Revert default checkpointing mode to EXACTLY_ONCE
- Use default for checkpoint timeout
- [BEAM-3089] Test default values of FlinkPipelineOptions
- Set default master url to [auto] 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by mxm** [comment]
CC @tweise @angoenka There are a couple of commits. The first commit fixes the parallelism issue that users reported with BEAM-3089. While fixing this I discovered that we don't have proper tests for the pipeline options and some of them are not set ideally.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 1]
```diff
@@ -156,9 +181,11 @@ public static StreamExecutionEnvironment createStreamExecutionEnvironment(
         throw new IllegalArgumentException("The checkpoint interval must be positive");
       }
       flinkStreamEnv.enableCheckpointing(checkpointInterval, options.getCheckpointingMode());
-      flinkStreamEnv
-          .getCheckpointConfig()
-          .setCheckpointTimeout(options.getCheckpointTimeoutMillis());
+      if (options.getCheckpointTimeoutMillis() != -1) {
```
Comment (angoenka) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: nit: < 0

File: runners/flink/src/test/resources/flink-conf.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+parallelism.default: 23
```
Comment (angoenka) [inline runners/flink/src/test/resources/flink-conf.yaml]: nit: new line

**Comment by mxm** [comment]
Retest this please

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 3]
```diff
@@ -56,12 +56,13 @@
...
           + " either be of the form \"host:port\" or one of the special values [local], "
           + "[collection] or [auto].")
+  @Default.String("[auto]")
   String getFlinkMaster();
 
   void setFlinkMaster(String value);
 
   @Description("The degree of parallelism to be used when distributing operations onto workers.")
-  @Default.InstanceFactory(DefaultParallelismFactory.class)
+  @Default.Integer(-1)
```
[the remark this answers is not in the document]
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: Updated.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 1 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: -1 in Flink means disabled. Anything else should be applied, even other negative numbers which will raise an exception.

**Comment by mxm** [comment]
Thanks for the review @angoenka.

**Comment by mxm** [comment]
@tweise As far as I see we have all the checkpointing related options already exposed. The problem is indeed that there might be more options that users want to configure. It could make sense to expose an interface to configure all Flink options.

**Pull request merged**